drivers/xen/gnttab.c

```diff
@@ -295,44 +299,67 @@ const char *gnttabop_error(int16_t status)
 	}
 }
 
+/* Picked from Linux implementation */
+#define LEGACY_MAX_GNT_FRAMES_SUPPORTED		4
+static unsigned long gnttab_get_max_frames(void)
+{
+	int ret;
+	struct gnttab_query_size q = {
+		.dom = DOMID_SELF,
+	};
+
+	ret = HYPERVISOR_grant_table_op(GNTTABOP_query_size, &q, 1);
+	if ((ret < 0) || (q.status != GNTST_okay)) {
+		return LEGACY_MAX_GNT_FRAMES_SUPPORTED;
+	}
+
+	return q.max_nr_frames;
+}
+
 static int gnttab_init(void)
 {
 	grant_ref_t gref;
 	struct xen_add_to_physmap xatp;
-	struct gnttab_setup_table setup;
-	xen_pfn_t frames[NR_GRANT_FRAMES];
 	int rc = 0, i;
+	unsigned long xen_max_grant_frames;
+	uintptr_t gnttab_base = DT_REG_ADDR_BY_IDX(DT_INST(0, xen_xen), 0);
+	mm_reg_t gnttab_reg;
+
+	xen_max_grant_frames = gnttab_get_max_frames();
+	if (xen_max_grant_frames < CONFIG_NR_GRANT_FRAMES) {
+		LOG_ERR("Xen max_grant_frames is less than CONFIG_NR_GRANT_FRAMES!");
+		k_panic();
+	}
 
 	/* Will be taken/given during gnt_refs allocation/release */
-	k_sem_init(&gnttab.sem, 0, NR_GRANT_ENTRIES - GNTTAB_NR_RESERVED_ENTRIES);
-
-	for (
-		gref = GNTTAB_NR_RESERVED_ENTRIES;
-		gref < NR_GRANT_ENTRIES;
-		gref++
-	    ) {
-		put_free_entry(gref);
+	k_sem_init(&gnttab.sem, NR_GRANT_ENTRIES - GNTTAB_NR_RESERVED_ENTRIES,
+		   NR_GRANT_ENTRIES - GNTTAB_NR_RESERVED_ENTRIES);
+
+	/* Initialize O(1) allocator, gnttab.gref_list[0] always shows first free entry */
+	for (gref = NR_GRANT_ENTRIES - 1; gref > GNTTAB_NR_RESERVED_ENTRIES; gref--) {
+		gnttab.gref_list[gref] = gnttab.gref_list[0];
+		gnttab.gref_list[0] = gref;
 	}
 
-	for (i = 0; i < NR_GRANT_FRAMES; i++) {
+	for (i = CONFIG_NR_GRANT_FRAMES; i; i--) {
 		xatp.domid = DOMID_SELF;
 		xatp.size = 0;
 		xatp.space = XENMAPSPACE_grant_table;
-		xatp.idx = i;
-		xatp.gpfn = xen_virt_to_gfn(Z_TOPLEVEL_ROM_NAME(grant_tables).phys_addr) + i;
+		xatp.idx = i - 1;
+		xatp.gpfn = xen_virt_to_gfn(gnttab_base) + (i - 1);
 		rc = HYPERVISOR_memory_op(XENMEM_add_to_physmap, &xatp);
 		__ASSERT(!rc, "add_to_physmap failed; status = %d\n", rc);
```

Comment:
AFAIU this assert and all above will work only in test builds: https://elixir.bootlin.com/zephyr/v4.0.0/source/subsys/debug/Kconfig#L249.

Is this expected, or should we use something else for assertions in any case?

Reply:
Yes, Zephyr assert exists only during testing build and compiles to nothing when testing is disabled. So, probably it should be changed to condition with some cleanup sequence. But once again, this lines were not touched in scope of this PR, so patches are welcomed :)

Reply:
I hope we can ensure this is addressed at some point to avoid potential issues. Perhaps it would be better to open an issue to track this.